File: hackerrank/medium/0000-forming-magic-square/python/solution.py
## solution.py

from typing import List

def forming_magic_square(s: List[List[int]]) -> int:
    n = 3

    def siamese(n: int) -> List[List[int]]:
        a = [[0] * n for _ in range(n)]
        i, j = 0, n // 2
        for num in range(1, n * n + 1):
            a[i][j] = num
            ni, nj = (i - 1) % n, (j + 1) % n
            if a[ni][nj] == 0:
                i, j = ni, nj
            else:
                i = (i + 1) % n
        return a

    def rotate90(m: List[List[int]]) -> List[List[int]]:
        return [[m[n - 1 - j][i] for j in range(n)] for i in range(n)]

    def mirror_vert(m: List[List[int]]) -> List[List[int]]:
        return m[::-1]

    base = siamese(n)
    cands = []
    cur = base
    for _ in range(4):
        cands.append(cur)
        cur = rotate90(cur)
    cur = mirror_vert(base)
    for _ in range(4):
        cands.append(cur)
        cur = rotate90(cur)

    best = 10**9
    for m in cands:
        cost = sum(abs(s[i][j] - m[i][j]) for i in range(n) for j in range(n))
        if cost < best:
            best = cost
    return best


# --- demo run ---
if __name__ == "__main__":
    print(forming_magic_square([[4,9,2],[3,5,7],[8,1,5]]))  # 1
    print(forming_magic_square([[4,8,2],[4,5,7],[6,1,6]]))  # 4

File: hackerrank/medium/0000-forming-magic-square/js/solution.mjs
// solution.mjs

export function formingMagicSquare(s) {
  const n = 3;

  function siamese(n) {
    const a = Array.from({ length: n }, () => Array(n).fill(0));
    let i = 0, j = Math.floor(n / 2);
    for (let num = 1; num <= n * n; num++) {
      a[i][j] = num;
      const ni = (i - 1 + n) % n;
      const nj = (j + 1) % n;
      if (a[ni][nj] === 0) { i = ni; j = nj; }
      else { i = (i + 1) % n; }
    }
    return a;
  }

  function rotate90(m) {
    const r = Array.from({ length: n }, () => Array(n).fill(0));
    for (let i = 0; i < n; i++) for (let j = 0; j < n; j++) r[j][n - 1 - i] = m[i][j];
    return r;
  }

  function mirrorVert(m) {
    return [...m].reverse();
  }

  const base = siamese(n);
  const cands = [];
  let cur = base;
  for (let k = 0; k < 4; k++) { cands.push(cur); cur = rotate90(cur); }
  cur = mirrorVert(base);
  for (let k = 0; k < 4; k++) { cands.push(cur); cur = rotate90(cur); }

  let best = Infinity;
  for (const m of cands) {
    let cost = 0;
    for (let i = 0; i < n; i++) for (let j = 0; j < n; j++) cost += Math.abs(s[i][j] - m[i][j]);
    if (cost < best) best = cost;
  }
  return best;
}

// --- demo run ---
console.log(formingMagicSquare([[4,9,2],[3,5,7],[8,1,5]])); // 1
console.log(formingMagicSquare([[4,8,2],[4,5,7],[6,1,6]])); // 4